bitcart/event_delivery.py

```diff
@@ -0,0 +1,174 @@
+import asyncio
+import logging
+import time  # noqa: F401: for sync generator
+from json import JSONDecodeError
+from typing import TYPE_CHECKING, Any, Callable, Dict, Iterable, Optional, Union
+
+if TYPE_CHECKING:
+    from providers.jsonrpcrequests import RPCProxy
+
+ASYNC = True
+
+
+webhook_available = True
+try:
+    if ASYNC:
+        from aiohttp import web
+    else:  # pragma: no cover
+        from flask import Flask, request
+except (ModuleNotFoundError, ImportError):  # pragma: no cover
+    webhook_available = False
+
+
+class EventDelivery:
+    server: "RPCProxy"
+    event_handlers: Dict[str, Callable]
+
+    def __init__(self) -> None:
+        self.webhook_available = webhook_available
+        if ASYNC:
+            self._configure_webhook = self._configure_webhook_async
+            self.handle_webhook = self.handle_webhook_async
+            self._start_webhook = self._start_webhook_async
+            self._loop = asyncio.get_event_loop()
+            self._loop.set_exception_handler(lambda loop, context: None)
+        else:  # pragma: no cover
+            self._configure_webhook = self._configure_webhook_sync
+            self.handle_webhook = self.handle_webhook_sync  # type: ignore
+            self._start_webhook = self._start_webhook_sync
+
+    def handle_webhook_sync(self) -> dict:
+        json = request.json
+        self.process_updates(json.get("updates", []), json.get("currency", "BTC"), json.get("wallet"))
+        return {}
+
+    async def handle_webhook_async(self, request: "web.Request") -> "web.Response":
+        try:
+            json = await request.json()
+        except JSONDecodeError:
+            return web.json_response({})
+        await self.process_updates(json.get("updates", []), json.get("currency", "BTC"), json.get("wallet"))
+        return web.json_response({})
+
+    async def process_updates(
+        self, updates: Iterable[dict], currency: Optional[str] = None, wallet: Optional[str] = None
+    ) -> None:
+        raise NotImplementedError()
+
+    def _configure_webhook_async(self) -> None:
+        self.webhook_app = web.Application()
+        self.webhook_app.router.add_post("/", self.handle_webhook)
+
+    def _configure_webhook_sync(self) -> None:
+        self.webhook_app = Flask(__name__)  # type: ignore
+        self.webhook_app.add_url_rule("/", "handle_webhook", self.handle_webhook, methods=["POST"])  # type: ignore
+
+    def check_webhook_support(self) -> None:
+        if not webhook_available:
+            raise ValueError("Webhook support not installed. Install it with pip install bitcart[webhook]")
+
+    async def _configure_notifications(self, autoconfigure: bool = True) -> None:
+        await self.server.subscribe(list(self.event_handlers.keys()))
+        if autoconfigure:
+            await self.server.configure_notifications("http://localhost:6000")
```

Comment:
do we want to hardcode  url here?

Reply:
I refactored that out from `btc.py`. It is the default port the server runs on, just for convenience. Users can pass `autoconfigure=False` and configure notifications themselves